Add LayerHostingContextManager

jyavenard opened the pull request:
#### f23a04c78df7376876bb1166c5f7f4e261f5b627
<pre>
Add LayerHostingContextManager
<a href="https://bugs.webkit.org/show_bug.cgi?id=299365">https://bugs.webkit.org/show_bug.cgi?id=299365</a>
<a href="https://rdar.apple.com/161161377">rdar://161161377</a>

Reviewed by Jer Noble.

We can then re-use the code without duplication.

No change in behaviour. Code just moved around.

* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::~RemoteMediaPlayerProxy):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h: Remove unused members.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxyConfiguration.h: Remove unused field
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxyConfiguration.serialization.in: Remove unused field
* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerFirstVideoFrameAvailable):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerRenderingModeChanged):
(WebKit::RemoteMediaPlayerProxy::requestHostingContext):
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeFenced):
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeIfPossible): Deleted.
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/LayerHostingContextManager.h: Added.
(WebKit::LayerHostingContextManager::videoLayerSize const):
* Source/WebKit/Platform/cocoa/LayerHostingContextManager.mm: Added.
(WebKit::LayerHostingContextManager::~LayerHostingContextManager):
(WebKit::LayerHostingContextManager::requestHostingContext):
(WebKit::LayerHostingContextManager::createHostingContextIfNeeded):
(WebKit::LayerHostingContextManager::setVideoLayerSizeFenced):
(WebKit::LayerHostingContextManager::setVideoLayerSizeIfPossible):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::createRemoteMediaPlayer):

Canonical link: <a href="https://commits.webkit.org/300560@main">https://commits.webkit.org/300560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80a62a7d78b33efbb71b85ac75ff39626cf4a124

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123087 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42801 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129743 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5b0f6dd4-7b2d-4252-a9f7-4661b8fd5fbf) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43525 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51396 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93561 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/66d04283-1c5b-4cf4-9249-ac5502efc6f2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126038 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34686 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110159 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74201 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ca274795-8270-446b-9158-6fb5dc662988) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73260 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104397 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28541 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132475 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50037 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38090 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102063 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50413 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106380 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25890 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47285 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25487 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49892 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55653 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52712 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
```diff
@@ -49,100 +49,43 @@
-    if (m_inlineLayerHostingContext)
-        m_inlineLayerHostingContext->setRootLayer(layer.get());
+    if (auto maybeHostingContext = layerHostingContextManager.createHostingContextIfNeeded(protectedPlayer()->platformLayer(), canShowWhileLocked))
+        protectedConnection()->send(Messages::MediaPlayerPrivateRemote::LayerHostingContextChanged(*maybeHostingContext, layerHostingContextManager.videoLayerSize()), m_id);
```
I don't think you need to name this variable `maybeHostingContext`. It's kind of implied with the `if()` check that it could fail to be created.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
```diff
@@ -473,6 +470,9 @@ class RemoteMediaPlayerProxy final
     const Ref<const Logger> m_logger;
 #endif
+#if PLATFORM(COCOA)
+    LayerHostingContextManager layerHostingContextManager;
```
Needs a `m_`. Can this be a `UniqueRef<>` so that we don't need to pull in `LayerHostingContextManager.h` into this header?

jernoble commented on Source/WebKit/Platform/cocoa/LayerHostingContextManager.h:
```diff
@@ -0,0 +1,57 @@
+    using LayerHostingContextCallback = CompletionHandler<void(WebCore::HostingContext)>;
+
+    void requestHostingContext(LayerHostingContextCallback&&);
+    std::optional<WebCore::HostingContext> createHostingContextIfNeeded(const PlatformLayerContainer&, bool canShowWhileLocked = false);
```
Do we need a default parameter here? It doesn't ever seem to be called without one.

jernoble reviewed (approved):
r=me with nits

jyavenard committed:
Add LayerHostingContextManager
https://bugs.webkit.org/show_bug.cgi?id=299365
rdar://161161377

Reviewed by Jer Noble.

We can then re-use the code without duplication.

No change in behaviour. Code just moved around.

* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::~RemoteMediaPlayerProxy):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h: Remove unused members.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxyConfiguration.h: Remove unused field
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxyConfiguration.serialization.in: Remove unused field
* Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerProxyCocoa.mm:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerFirstVideoFrameAvailable):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerRenderingModeChanged):
(WebKit::RemoteMediaPlayerProxy::requestHostingContext):
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeFenced):
(WebKit::RemoteMediaPlayerProxy::setVideoLayerSizeIfPossible): Deleted.
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/LayerHostingContextManager.h: Added.
(WebKit::LayerHostingContextManager::videoLayerSize const):
* Source/WebKit/Platform/cocoa/LayerHostingContextManager.mm: Added.
(WebKit::LayerHostingContextManager::~LayerHostingContextManager):
(WebKit::LayerHostingContextManager::requestHostingContext):
(WebKit::LayerHostingContextManager::createHostingContextIfNeeded):
(WebKit::LayerHostingContextManager::setVideoLayerSizeFenced):
(WebKit::LayerHostingContextManager::setVideoLayerSizeIfPossible):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.cpp:
(WebKit::RemoteMediaPlayerManager::createRemoteMediaPlayer):

Canonical link: https://commits.webkit.org/300560@main
